scipy/_lib/src/_test_ccallback.c

```diff
@@ -0,0 +1,363 @@
+/*
+ * Test code for ccallback.h
+ *
+ * This also is an internal "best-practices" code example on how to write
+ * low-level callback code.
+ *
+ * The *thunk_simple* function shows how to write the wrapper callback
+ * function that dispatches to user-provided code (written in Python, C, Cython etc.).
+ *
+ * The *call_simple* function shows how to setup and teardown the ccallback_t
+ * data structure.
+ *
+ * The *call_nodata* and *thunk_nodata* show what to do when you need a
+ * callback function for some code where it's not possible to pass along a
+ * custom data pointer.
+ *
+ * The *call_nonlocal* and *thunk_nonlocal* show how to use setjmp/longjmp
+ * to obtain a nonlocal return on error conditions, in cases where there's no
+ * mechanism to interrupt computation. Note that this is the last-resort option,
+ * and only safe if there is no memory allocation between setjmp/longjmp (or you
+ * need to add additonal cleanup yourself).
+ *
+ */
+
+#include <setjmp.h>
+#include <Python.h>
+
+#include "ccallback.h"
+
+
+#define ERROR_VALUE 2
+
+/* Thunks for the different cases considered */
+
+static double test_thunk_simple(double a, int *error_flag, void *data)
+{
+    ccallback_t *callback = (ccallback_t*)data;
+    double result = 0;
+    int error = 0;
+
+    if (callback->py_function) {
+        PyGILState_STATE state = PyGILState_Ensure();
+        PyObject *res, *res2;
+
+        res = PyObject_CallFunction(callback->py_function, "d", a);
+
+        if (res == NULL) {
+            error = 1;
+        }
+        else {
+            res2 = PyNumber_Float(res);
+            if (res2 == NULL) {
+                error = 1;
+            }
+            else {
+                result = PyFloat_AsDouble(res2);
+                if (PyErr_Occurred()) {
+                    error = 1;
+                }
+                Py_DECREF(res2);
+            }
+            Py_DECREF(res);
+        }
+
+        PyGILState_Release(state);
+    }
+    else {
+        if (callback->signature_index == 0) {
+            result = ((double(*)(double, int *, void *))callback->c_function)(
+                a, &error, callback->user_data);
+        }
+        else {
+            result = ((double(*)(double, double, int *, void *))callback->c_function)(
+                a, 0.0, &error, callback->user_data);
+        }
+    }
+
+    if (error) {
+        *error_flag = 1;
```

Comment:
I'm a bit surprised that `*error_flag` isn't set to 0 explicitly otherwise. Is it best practice?